File: core/src/exporter.rs
use crate::palette::DesignSystemTokens;
use serde_json::{json, Value};

pub enum ExportFormat {
    Css,
    Scss,
    TailwindV3,
    TailwindV4,
    FigmaTokens,
    StyleDictionary,
    Swift,
    AndroidXml,
}

pub struct TokenExporter;

impl TokenExporter {
    pub fn export_css(tokens: &DesignSystemTokens) -> String {
        let mut out = String::new();
        out.push_str(":root {\n");
        for (k, v) in &tokens.light {
            out.push_str(&format!("  {}: {};\n", k, v));
        }
        out.push_str("}\n\n[data-theme=\"dark\"], .dark {\n");
        for (k, v) in &tokens.dark {
            out.push_str(&format!("  {}: {};\n", k, v));
        }
        out.push_str("}\n");
        out
    }

    pub fn export_scss(tokens: &DesignSystemTokens) -> String {
        let mut out = String::new();
        out.push_str("// Light Theme\n");
        for (k, v) in &tokens.light {
            let var_name = k.trim_start_matches("--");
            out.push_str(&format!("${}: {};\n", var_name, v));
        }
        out.push_str("\n// Dark Theme\n");
        for (k, v) in &tokens.dark {
            let var_name = k.trim_start_matches("--");
            out.push_str(&format!("${}-dark: {};\n", var_name, v));
        }
        out
    }

    pub fn export_tailwind_v3(tokens: &DesignSystemTokens) -> String {
        let mut colors = String::new();
        for (k, _) in &tokens.light {
            let name = k.trim_start_matches("--");
            let js_name = name.replace('-', "_");
            colors.push_str(&format!("        '{}': 'var({})',\n", js_name, k));
        }

        format!(
            "module.exports = {{\n  theme: {{\n    extend: {{\n      colors: {{\n{}      }}\n    }}\n  }}\n}};\n",
            colors
        )
    }

    pub fn export_tailwind_v4(tokens: &DesignSystemTokens) -> String {
        let mut out = String::new();
        out.push_str("@theme {\n");
        for (k, v) in &tokens.light {
            let name = k.trim_start_matches("--");
            out.push_str(&format!("  --color-{}: {};\n", name, v));
        }
        out.push_str("}\n");
        out
    }

    pub fn export_figma_tokens(tokens: &DesignSystemTokens) -> String {
        let mut light_map = serde_json::Map::new();
        for (k, v) in &tokens.light {
            let name = k.trim_start_matches("--").to_string();
            light_map.insert(
                name,
                json!({
                    "$value": v,
                    "$type": "color"
                }),
            );
        }

        let mut dark_map = serde_json::Map::new();
        for (k, v) in &tokens.dark {
            let name = k.trim_start_matches("--").to_string();
            dark_map.insert(
                name,
                json!({
                    "$value": v,
                    "$type": "color"
                }),
            );
        }

        let root = json!({
            "global": {
                "light": light_map,
                "dark": dark_map
            }
        });

        serde_json::to_string_pretty(&root).unwrap_or_default()
    }

    pub fn export_style_dictionary(tokens: &DesignSystemTokens) -> String {
        let mut color_tree = serde_json::Map::new();
        for (k, v) in &tokens.light {
            let name = k.trim_start_matches("--").to_string();
            color_tree.insert(
                name,
                json!({
                    "value": v,
                    "type": "color"
                }),
            );
        }

        let root = json!({
            "color": color_tree
        });

        serde_json::to_string_pretty(&root).unwrap_or_default()
    }

    pub fn export_swift(tokens: &DesignSystemTokens) -> String {
        let mut out = String::new();
        out.push_str("import SwiftUI\n\npublic enum AppTheme {\n    public enum Colors {\n");
        for (k, v) in &tokens.light {
            let name = k.trim_start_matches("--").replace('-', "_");
            let clean_hex = v.trim_start_matches('#');
            if clean_hex.len() == 6 {
                let r = u8::from_str_radix(&clean_hex[0..2], 16).unwrap_or(0) as f64 / 255.0;
                let g = u8::from_str_radix(&clean_hex[2..4], 16).unwrap_or(0) as f64 / 255.0;
                let b = u8::from_str_radix(&clean_hex[4..6], 16).unwrap_or(0) as f64 / 255.0;
                out.push_str(&format!(
                    "        public static let {} = Color(red: {:.3}, green: {:.3}, blue: {:.3})\n",
                    name, r, g, b
                ));
            }
        }
        out.push_str("    }\n}\n");
        out
    }

    pub fn export_android_xml(tokens: &DesignSystemTokens) -> String {
        let mut out = String::new();
        out.push_str("<?xml version=\"1.0\" encoding=\"utf-8\"?>\n<resources>\n");
        for (k, v) in &tokens.light {
            let name = k.trim_start_matches("--").replace('-', "_");
            out.push_str(&format!("    <color name=\"{}\">{}</color>\n", name, v));
        }
        out.push_str("</resources>\n");
        out
    }

    pub fn export(tokens: &DesignSystemTokens, format: ExportFormat) -> String {
        match format {
            ExportFormat::Css => Self::export_css(tokens),
            ExportFormat::Scss => Self::export_scss(tokens),
            ExportFormat::TailwindV3 => Self::export_tailwind_v3(tokens),
            ExportFormat::TailwindV4 => Self::export_tailwind_v4(tokens),
            ExportFormat::FigmaTokens => Self::export_figma_tokens(tokens),
            ExportFormat::StyleDictionary => Self::export_style_dictionary(tokens),
            ExportFormat::Swift => Self::export_swift(tokens),
            ExportFormat::AndroidXml => Self::export_android_xml(tokens),
        }
    }
}

File: wasm/src/exporter.ts
import { DesignSystemTokens } from './types.js';

export type ExportFormat =
  | 'css'
  | 'scss'
  | 'tailwind-v3'
  | 'tailwind-v4'
  | 'figma-tokens'
  | 'style-dictionary'
  | 'swift'
  | 'android-xml';

export function exportCss(tokens: DesignSystemTokens): string {
  let out = ':root {\n';
  for (const [k, v] of Object.entries(tokens.light)) {
    out += `  ${k}: ${v};\n`;
  }
  out += '}\n\n[data-theme="dark"], .dark {\n';
  for (const [k, v] of Object.entries(tokens.dark)) {
    out += `  ${k}: ${v};\n`;
  }
  out += '}\n';
  return out;
}

export function exportScss(tokens: DesignSystemTokens): string {
  let out = '// Light Theme\n';
  for (const [k, v] of Object.entries(tokens.light)) {
    const varName = k.replace(/^--/, '');
    out += `$${varName}: ${v};\n`;
  }
  out += '\n// Dark Theme\n';
  for (const [k, v] of Object.entries(tokens.dark)) {
    const varName = k.replace(/^--/, '');
    out += `$${varName}-dark: ${v};\n`;
  }
  return out;
}

export function exportTailwindV3(tokens: DesignSystemTokens): string {
  let colors = '';
  for (const k of Object.keys(tokens.light)) {
    const name = k.replace(/^--/, '');
    const jsName = name.replace(/-/g, '_');
    colors += `        '${jsName}': 'var(${k})',\n`;
  }

  return `module.exports = {
  theme: {
    extend: {
      colors: {
${colors}      }
    }
  }
};
`;
}

export function exportTailwindV4(tokens: DesignSystemTokens): string {
  let out = '@theme {\n';
  for (const [k, v] of Object.entries(tokens.light)) {
    const name = k.replace(/^--/, '');
    out += `  --color-${name}: ${v};\n`;
  }
  out += '}\n';
  return out;
}

export function exportFigmaTokens(tokens: DesignSystemTokens): string {
  const lightMap: Record<string, { $value: string; $type: string }> = {};
  for (const [k, v] of Object.entries(tokens.light)) {
    const name = k.replace(/^--/, '');
    lightMap[name] = {
      $value: v,
      $type: 'color',
    };
  }

  const darkMap: Record<string, { $value: string; $type: string }> = {};
  for (const [k, v] of Object.entries(tokens.dark)) {
    const name = k.replace(/^--/, '');
    darkMap[name] = {
      $value: v,
      $type: 'color',
    };
  }

  const root = {
    global: {
      light: lightMap,
      dark: darkMap,
    },
  };

  return JSON.stringify(root, null, 2);
}

export function exportStyleDictionary(tokens: DesignSystemTokens): string {
  const colorTree: Record<string, { value: string; type: string }> = {};
  for (const [k, v] of Object.entries(tokens.light)) {
    const name = k.replace(/^--/, '');
    colorTree[name] = {
      value: v,
      type: 'color',
    };
  }

  const root = {
    color: colorTree,
  };

  return JSON.stringify(root, null, 2);
}

export function exportSwift(tokens: DesignSystemTokens): string {
  let out = 'import SwiftUI\n\npublic enum AppTheme {\n    public enum Colors {\n';
  for (const [k, v] of Object.entries(tokens.light)) {
    const name = k.replace(/^--/, '').replace(/-/g, '_');
    const cleanHex = v.replace(/^#/, '');
    if (cleanHex.length === 6) {
      const r = (parseInt(cleanHex.substring(0, 2), 16) / 255.0).toFixed(3);
      const g = (parseInt(cleanHex.substring(2, 4), 16) / 255.0).toFixed(3);
      const b = (parseInt(cleanHex.substring(4, 6), 16) / 255.0).toFixed(3);
      out += `        public static let ${name} = Color(red: ${r}, green: ${g}, blue: ${b})\n`;
    }
  }
  out += '    }\n}\n';
  return out;
}

export function exportAndroidXml(tokens: DesignSystemTokens): string {
  let out = '<?xml version="1.0" encoding="utf-8"?>\n<resources>\n';
  for (const [k, v] of Object.entries(tokens.light)) {
    const name = k.replace(/^--/, '').replace(/-/g, '_');
    out += `    <color name="${name}">${v}</color>\n`;
  }
  out += '</resources>\n';
  return out;
}

export function exportTokens(tokens: DesignSystemTokens, format: ExportFormat): string {
  switch (format) {
    case 'css':
      return exportCss(tokens);
    case 'scss':
      return exportScss(tokens);
    case 'tailwind-v3':
      return exportTailwindV3(tokens);
    case 'tailwind-v4':
      return exportTailwindV4(tokens);
    case 'figma-tokens':
      return exportFigmaTokens(tokens);
    case 'style-dictionary':
      return exportStyleDictionary(tokens);
    case 'swift':
      return exportSwift(tokens);
    case 'android-xml':
      return exportAndroidXml(tokens);
    default:
      return exportCss(tokens);
  }
}
